source/module_hsolver/module_pexsi/dist_matrix_transformer.cpp

```diff
@@ -0,0 +1,765 @@
+                    minimum_process = i;
+                }
+            }
+        }
+        if (minimum_process >= 0) // find it!
+        {
+            ++displacement_process[minimum_process];
+            --size_process[minimum_process];
+        }
+        pre_position = minimum_position;
+        pre_process = minimum_process;
+        return minimum_position;
+    }
+}
+
+inline void DistMatrixTransformer::buildCCSParameter(const int size,
+                                                     const int nprocs,
+                                                     std::vector<int> size_process,
+                                                     std::vector<int> displacement_process,
+                                                     const int* position_index,
+                                                     DistCCSMatrix& DST_Matrix,
+                                                     int* buffer2ccsIndex)
+{
+    // find the minimum one from left buffer index
+    if (DST_Matrix.get_nnzlocal() <= 0)
+        return;
+
+    int pre_col = -1;
+    int nnz_now = 0;
+    int p_mini;
+    p_mini = MinimumIndexPosition(true, nprocs, &size_process[0], &displacement_process[0], position_index);
+    while (p_mini >= 0)
+    {
+        int index_mini = position_index[p_mini];
+        int col_mini = index_mini / DST_Matrix.get_size(); //-DST_Matrix.firstCol;
+        int row_mini = index_mini % DST_Matrix.get_size();
+        if (col_mini > pre_col) // a new column starts, column pointer is a 1-based array
+        {
+            pre_col = col_mini;
+            DST_Matrix.get_colptr_local()[col_mini] = nnz_now + 1;
+        }
+        DST_Matrix.get_rowind_local()[nnz_now] = row_mini + 1; // setup row index array, which is also 1-based
+        // copy data from buffer to M, be careful M is a 0-based array
+        buffer2ccsIndex[nnz_now] = p_mini;
+        ++nnz_now;
+        p_mini = MinimumIndexPosition(false, nprocs, &size_process[0], &displacement_process[0], position_index);
+    }
+    // The last element of colptrLocal is nnzLocal+1
+    DST_Matrix.get_colptr_local()[DST_Matrix.get_numcol_local()] = nnz_now + 1;
+}
+
+inline void DistMatrixTransformer::buffer2CCSvalue(int nnzLocal,
+                                                   int* buffer2ccsIndex,
+                                                   double* buffer,
+                                                   double* nzvalLocal)
+{
+    for (int i = 0; i < nnzLocal; ++i)
+    {
+        nzvalLocal[i] = buffer[buffer2ccsIndex[i]];
+    }
+}
+inline void DistMatrixTransformer::countMatrixDistribution(int N, double* A, std::map<int, int>& P)
+{
+    for (int i = 0; i < N; ++i)
+    {
+        int key;
+        if (fabs(A[i] < 1e-31))
+            key = -100;
+        else
+            key = floor(log10(fabs(A[i])));
+        ++P[key];
+    }
+}
+
+// find out the index of non-zero elements
+inline int DistMatrixTransformer::getNonZeroIndex(char layout,
```

Comment:
you can set layout to const char&, but I am also a little bit confused about why there is `layout == 'r' || layout == 'R'`, is it possible to avoid the value of layout to be uppercase or lowercase?

Reply:
it seems that the layout is always 'c', which was passed to here in the call of `simplePEXSI`, so these can always be lowercase. I'll fix this